Fix a few bytes/unicode issues in conch script

rodrigc opened the pull request:
https://twistedmatrix.com/trac/ticket/9344

Fix a few bytes/unicode issues in the conch script, and fix a few twistedchecker style issues.

Fix one case where if you do:

```
conch localhost "nonexistentcommand"
```

instead of getting this:

```
Unhandled Error
               Traceback (most recent call last):
                                                   File "/Users/crodrigues/twisted4/src/twisted/python/context.py", line 122, in callWithContext
                                                                                                                                                    return self.currentContext().callWithContext(ctx, func, *args, **kw)
                                                             File "/Users/crodrigues/twisted4/src/twisted/python/context.py", line 85, in callWithContext
                                                                                                                                                             return func(*args,**kw)
                         File "/Users/crodrigues/twisted4/src/twisted/conch/ssh/service.py", line 45, in packetReceived
                                                                                                                           return f(packet)
                                                                                                                                             File "/Users/crodrigues/twisted4/src/twisted/conch/ssh/connection.py", line 273, in ssh_CHANNEL_EXTENDED_DATA
                                                                                                 log.callWithLogger(channel, channel.extReceived, typeCode, data)
    --- <exception caught here> ---
                                     File "/Users/crodrigues/twisted4/src/twisted/python/log.py", line 103, in callWithLogger
                                                                                                                                 return callWithContext({"system": lp}, func, *args, **kw)
                               File "/Users/crodrigues/twisted4/src/twisted/python/log.py", line 86, in callWithContext
                                                                                                                           return context.call({ILogContext: newCtx}, func, *args, **kw)
                             File "/Users/crodrigues/twisted4/src/twisted/python/context.py", line 122, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/Users/crodrigues/twisted4/src/twisted/python/context.py", line 85, in callWithContext
    return func(*args,**kw)
  File "/Users/crodrigues/twisted4/src/twisted/conch/scripts/conch.py", line 468, in extReceived
    sys.stderr.write(data)
builtins.TypeError: write() argument must be str, not bytes
```

You get something like:

```
zsh:1: command not found: nonexistentcommand
```

rodrigc committed:
Switch to format strings

rodrigc committed:
Fix some whitespace issues flagged by twistedchecker

rodrigc committed:
On Python 3, write bytes to sys.stderr.buffer, not sys.stderr.

rodrigc committed:
Add newsfragment

rodrigc committed:
handleInput() is an implementation of dataReceived() and must deal with bytes

rodrigc committed:
Fix whitespace

rodrigc committed:
Clean up lambda to appease twistedchecker

rodrigc committed:
Add a __next__ method for using as iterator on Python 3.

rodrigc commented:
@wiml can you try this patch?

wiml commented on src/twisted/conch/ssh/filetransfer.py:
```diff
@@ -816,35 +816,45 @@ def setAttrs(self, attrs):
         d.addErrback(self._ebReadDir)
         return d
 
+    next = __next__
```
`_ebReadDir` modifies the `next` attribute at the end of the listing; it will have to either modify `__next__` as well or one of these should call the other

wiml commented:
I added a comment to the diff.

While you're looking at this, could you take a look at fixing [#7322](https://twistedmatrix.com/trac/ticket/7322) as well? That's the other bug I currently have to work around in order to use the directory iterator, and it seems like it should also have a nearly trivial fix.

rodrigc committed:
Modify __next__ in _ebReaddir

rodrigc commented:
Please look at the updated patch, where I deal with __next__ and next.

Do you have a small testcase that we can use to trigger  [#7322](https://twistedmatrix.com/trac/ticket/7322)?

wiml commented:
Here's a test case you can add to `test_filetransfer.py`/`OurServerOurClientTests` to demonstrate both [#7322](https://twistedmatrix.com/trac/ticket/7322) and [#9349](https://twistedmatrix.com/trac/ticket/9349):

```python
    @defer.inlineCallbacks
    def testOpenDirectoryIterator(self):
        """
        Check that the object returned by
        L{filetransfer.FileTransferClient.openDirectory} can be used
        as an iterator.
        """

        # This function is a little more complicated than it would be
        # normally, since we need to call _emptyBuffers() after
        # creating any SSH-related Deferreds, but before waiting on
        # them via yield.
        
        d = self.client.openDirectory(b'')
        self._emptyBuffers()
        openDir = yield d
        
        filenames = set()
        try:
            for f in openDir:
                self._emptyBuffers()
                (filename, _, fileattrs) = yield f
                filenames.add(filename)
        finally:
            d = openDir.close()
            self._emptyBuffers()
            yield d
            
        self._emptyBuffers()
        
        self.assertEqual(filenames,
                         set([b'.testHiddenFile', b'testDirectory',
                              b'testRemoveFile', b'testRenameFile',
                              b'testfile1']))

```

rodrigc committed:
Add test for the iterator returned by openDirectory().

rodrigc committed:
Fix whitespace

rodrigc committed:
Use next()

rodrigc committed:
Convert EOFError to StopIteration so that openDirectory iterator works properly

rodrigc committed:
Merge branch 'trunk' into 9344-rodrigc-conch-py3

rodrigc commented:
@wiml can you try this updated patch, which addresses [#7322](https://twistedmatrix.com/trac/ticket/7322)? also?

wiml commented on src/twisted/conch/ssh/filetransfer.py:
```diff
@@ -856,12 +856,8 @@ def _cbReadDir(self, names):
-        def _():
-            raise StopIteration
-        self.__next__ = _
-        self.next = self.__next__
```
This change removes the monkeypatching of `next`/`__next__` (which I think is good, stylistically) but it changes the behavior of the iterator. It no longer remembers that it has reached EOF, and calling `next` again will cause a protocol violation (issuing `READDIR` on a closed handle). Python iterators [are required](https://docs.python.org/3/library/stdtypes.html#iterator.__next__) to allow calling `next` after StopIteration (and should continue raising StopIteration).

wiml commented:
I think the last change (ba968f180) breaks the iterator protocol. I have an alternate suggestion, I'll post it as soon as I have a chance to test it.

wiml committed:
Raise StopIteration from the directory iterator when the server
returns EOF, and make sure it's safe to continue calling next even
after that (as required by the Python iterator protocol).

wiml commented:
Try this fix instead: wiml/twisted@ad321dcc0

rodrigc committed:
Merge branch 'trunk' into 9344-rodrigc-conch-py3

rodrigc commented:
@wiml I merged your change, and it looks good.  Is this OK to merge?

wiml commented:
I might not be using github's viewer correctly, but it doesn't look like this PR contains wiml/twisted@ad321dcc0 or fixes the problem that it is supposed to fix — the [iterator protocol requires](https://docs.python.org/3/library/stdtypes.html#iterator-types) that you can continue calling `next` on a finished iterator and keep safely getting a `StopIteration`.

rodrigc committed:
Merge branch 'trunk' into 9344-rodrigc-conch-py3

rodrigc commented:
@wiml can you take another look?  I've merged your patch.

wiml commented:
@rodrigc  - That looks good to me and works in my use case.

rodrigc committed:
Replace networkString() with ensureBytes()

rodrigc committed:
Merge branch 'trunk' into 9344-rodrigc-conch-py3

rodrigc committed:
Handle and dirObj must be bytes to work on Python 3

rodrigc committed:
Rename variables to appease twistedchecker

rodrigc committed:
Fix twistedchecker errors

rodrigc committed:
Fix twistedchecker warnings

rodrigc committed:
Fix twistedchecker errors

rodrigc committed:
Fix pycodestyle errors
